### ISSUE TITLE
Toucan Memory Leaks

### ISSUE THREAD

**Issue by DrewKiino** [gavinbunney/Toucan#29]
Hi, I use your library to resize my images in my tableview and I noticed that the memory keeps increasing in my app when I use instruments. It shows your library and I checked out your code and I noticed that theres not memory management handling in your closures. It shows that on line 598 on the Toucan class reaching 200+mb's in the closure. Do you mind if I make a pull request and do a fix on it? I'm using your library for a production app

**Comment by DrewKiino** [comment]
I think I figured out the problem. It's the function drawImageWithClosure where the size is being passed in from the outside class, where UIGraphicsBeginImageContextWithOptions takes in the size argument. I defered the reference to a variable inside the function and the leak went away. But I would double check.

**Comment by gavinbunney** [comment]
Thanks Drew, have you had a chance to confirm the fix? Happy for you to submit a pull request with it :+1:

**Comment by DrewKiino** [comment]
Yeah, it is confirmed. Your library is working nicely! I kind of modified your code to take in optionals because it was difficult unwrapping sizes when I used your library in background threads. I'll go over it a bit and clean it up then ill do a pull request, let me know if its acceptable. 

    On Sunday, January 3, 2016 5:44 PM, Gavin Bunney <notifications@github.com> wrote:
 

 Thanks Drew, have you had a chance to confirm the fix? Happy for you to submit a pull request with it  —

**Comment by JoeFerrucci** [comment]
@gavinbunney @DrewKiino This is an issue for me. I am using `Version 0.5.0`. What's the status on this issue?

**Comment by gavinbunney** [comment]
Hi @JoeFerrucci - I haven't looked at this issue in much detail; is this still an issue with Xcode 8.1/Swift 3? I would hope that it was more of an issue on Apple's side since the `UIGraphicsBeginImageContextWithOptions` argument but if it still is I'll take a look

**Comment by abellee** [comment]
is this been fixed?

**Comment by nevinjethmalani** [comment]
I am having issues with this as well. Do you have this on your timeline. As soon as we put in 5 or more larger sized images to process on the background thread, it crashes. @DrewKiino were you able to fix this issue?

**Comment by nevinjethmalani** [comment]
When I tested this using the Instruments tool there was a huge spike in memory usage as Toucan was being used. @DrewKiino and @gavinbunney we would love help with this as this is causing a crash/OOM in our app anytime we import a large set of images and resize them.

**Comment by kean** [comment]
Hey, @nevinjethmalani. Have you tried using [autorelease pools](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmAutoreleasePools.html#//apple_ref/doc/uid/20000047-SW2)?

**Comment by nevinjethmalani** [comment]
No how would we do that using swift? 

Here is the code that we use is swift `newSelectedimage = Toucan(image: newSelectedimage).resize(CGSize(width: widthInPixels, height: (widthInPixels*5)/4), fitMode: Toucan.Resize.FitMode.crop).image`

widthInPixels is just a constant and newSelectedimage is the image.

**Comment by kean** [comment]
```swift
newSelectedimage = autoreleasepool {
    Toucan(image: newSelectedimage).resize(CGSize(width: widthInPixels, height: (widthInPixels*5)/4), fitMode: Toucan.Resize.FitMode.crop).image
}
```

**Comment by nevinjethmalani** [comment]
Unfortunately that didn't work. I am still getting the crash/OOM when doing it with a group of images.

**Comment by nevinjethmalani** [comment]
@kean Unfortunately that didn't work. I am still getting the crash/OOM when doing it with a group of images. But thanks a lot for the quick response. Do you have any other thoughts on what might help?

**Comment by kean** [comment]
[Profile memory usage](https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/InstrumentsUserGuide/CommonMemoryProblems.html#//apple_ref/doc/uid/TP40004652-CH91-SW1), find out what takes some much memory and/or leaks, fix it.

**Comment by porzal** [comment]
I am not sure if this issue is related to this, I have 7mb large pano image and your library can't handle it and my app crashes during resizing the image.

**Comment by OO-E** [comment]
+1

**Comment by alexanderkhitev** [comment]
Any updates?

**Comment by DrewKiino** [comment]
I have a fix for this somewhere in a past project, going to try to see if I can merge the latest then PR back in.

**Comment by DrewKiino** [comment]
here's a fix - https://github.com/gavinbunney/Toucan/pull/68

the memory leaks occurred when `drawImageWithClosure` was called but the callback lost reference to the caller (ex: the image view that needed Toucan was deallocated). `UIGraphicsEndImageContext` was never called since `closure(size, context)` never got past the callback stack.

I stress tested this with 5120 by 3200 images at 100 cells using SDWebImage and no caching so it continually processed these large images. I only tested `maskWithEllipse` but I added the same fix to all the other functions.
